cuqi/distribution/_distribution.py

```diff
@@ -125,8 +126,60 @@ def geometry(self,value):
         else:
             raise TypeError("The attribute 'geometry' should be of type 'int' or 'cuqi.geometry.Geometry', or None.")
 
+    def logd(self, *args, **kwargs):
+        """  Evaluate the un-normalized log density function of the distribution.
+
+        The log density function is equal to the log probability density function (logpdf) of a distribution
+        plus an additive constant.
+
+        It is possible to pass conditioning variables as arguments to this function in addition to the parameters of the distribution.
+
+        All distributions are required to implement an un-normalized log density function, but not required
+        implement the log probability density function (logpdf). For MCMC sampling, the log density function is
+        used to sample from the distribution for efficient sampling.
+        
+        """
+
+        # Get the (potential) conditioning variables
+        cond_vars = self.get_conditioning_variables()
+
+        # If distribution is conditional, we first condition before evaluating the log density
+        if len(cond_vars) > 0:
+
+            if len(args) > 0:
+                raise ValueError(f"{self.logd.__qualname__}: Positional arguments on conditional distributions are not supported in logd evaluation (yet)")
+
+            # Check if all conditioning variables are specified
+            if not all([key in kwargs for key in cond_vars]):
+                raise ValueError(f"{self.logd.__qualname__}: To evaluate the log density all conditioning variables must be specified. Conditioning variables are: {cond_vars}")
+
+            # Extract exactly the conditioning variables from kwargs
+            cond_kwargs = {key: kwargs[key] for key in cond_vars}
+
+            # Extract any remaining variables from kwargs
+            non_cond_kwargs = {key: kwargs[key] for key in kwargs if key not in cond_vars}
+
+            # Condition the distribution on the conditioning variables
+            new_dist = self(**cond_kwargs)
+
+            # Now evaluate the log density of the fully specified distribution
+            return new_dist.logd(*args, **non_cond_kwargs)
+
+        # Not conditional distribution, simply evaluate log density directly
+        else:
+            return super().logd(*args, **kwargs)
+        
+    def _logd(self, *args):
+        return self.logpdf(*args) # Currently all distributions implement logpdf so we simply call this method.
+
     @abstractmethod
     def logpdf(self,x):
+        """ Evaluate the log probability density function of the distribution.
```

Comment:
I can't see much about the constant that separates logd and logpdf. It is briefly mentioned under the logd. I would maybe expect that density would have a normalization constant attribute that could be queried (default zero probably) and that the logpdf implementation would return logd plus the constant in the base function, with that the concrete classes provide specific implementation of logpdf. Or did we agree to postpone this?

Reply:
This is postoned to #16. Currently all distributions implement logpdf and logd calls logpdf so there is no constant :)